### PR TITLE
add stream::count

### PR DESCRIPTION
**Pull request by starsheriff**
ref #129

### CODE REVIEW

File: src/stream/stream/count.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    pin_utils::unsafe_unpinned!(count: usize);
+
+    pub(crate) fn new(stream: S) -> Self {
+        CountFuture { stream, count: 0 }
+    }
+}
+
+impl<S> Future for CountFuture<S>
+where
+    S: Sized + Stream,
```
Comment (stjepang) [inline src/stream/stream/count.rs]: I believe `S: Sized` is redundant here, i.e. `S` is assumed to be `Sized` by default unless we opt-out with `S: ?Sized`.

```suggestion
    S: Stream,
```

Comment (starsheriff) [reply src/stream/stream/count.rs]: Actually I don't remember why I used it in the first place. I think the compiler was complaining about it at some point. I actually don't see anymore why it should be `Sized` in the first place.

Will remove it.

File: src/stream/stream/mod.rs [hunk 2]
```diff
@@ -1392,6 +1394,33 @@ extension_trait! {
...
+
+            assert_eq!(s1.count().await, 1);
+            assert_eq!(s2.count().await, 3);
+            #
+            # }) }
+            ```
+        "#]
+        fn count(self) -> impl Future<Output = Ordering> [CountFuture<Self>]
+        where
+            Self: Sized + Stream,
```
Comment (k-nasa) [inline src/stream/stream/mod.rs]: Do you need this 'Stream'?

Comment (starsheriff) [reply src/stream/stream/mod.rs]: Actually not, your're right.

**Comment by yoshuawuyts** [comment]
Filed https://github.com/async-rs/async-std/pull/401 to fix CI. Let's re-run it after we merge that. Thanks!

File: src/stream/stream/count.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#[doc(hidden)]
+#[allow(missing_debug_implementations)]
+pub struct CountFuture<S> {
+    stream: S,
+    count: usize,
+}
+
+impl<S> CountFuture<S> {
+    pin_utils::unsafe_pinned!(stream: S);
+    pin_utils::unsafe_unpinned!(count: usize);
```
Comment (stjepang) [inline src/stream/stream/count.rs]: Can we rebase this branch and use `pin_project!` here?

Comment (starsheriff) [reply src/stream/stream/count.rs]: Yes, can do that. Sorry for the late reply, I was quite busy the last weeks.

**Comment by starsheriff** [comment]
so, changed to using `pin_project`. Guess it is ready now.

**Comment by starsheriff** [comment]
@stjepang, can you have a look?

File: src/stream/stream/mod.rs [hunk 4]
```diff
@@ -1802,6 +1804,33 @@ extension_trait! {
...
+            let s1 = stream::from_iter(vec![0]);
+            let s2 = stream::from_iter(vec![1, 2, 3]);
+
+            assert_eq!(s1.count().await, 1);
+            assert_eq!(s2.count().await, 3);
+            #
+            # }) }
+            ```
+        "#]
+        fn count(self) -> impl Future<Output = Ordering> [CountFuture<Self>]
```
Comment (k-nasa) [inline src/stream/stream/mod.rs]: https://github.com/async-rs/async-std/pull/368/files#diff-14e289ac8d84db1191a59c7117c5a476R29
Here is usize.

Why did you choose Ordering here?

```suggestion
        fn count(self) -> impl Future<Output = usize> [CountFuture<Self>]
```

Comment (starsheriff) [reply src/stream/stream/mod.rs]: I tried to keep it as generic as possible first, but I guess there is no point in it.

**Pull request merged**